merqube_client_lib/api_client/base.py

```diff
@@ -124,45 +180,73 @@ def index_post_model_from_existing(
 
         return model
 
-    def create_index(self, index_def: IndexDefinitionPost) -> dict[str, str]:
+    @_name_or_id
+    def delete_index(self, index_name: str | None = None, index_id: str | None = None) -> HTTPResJson:
         """
-        Create an index
-        Returns a dictionary containing the id of the index and its related securities (index, intraday_index)
-
-        TODO: examples and index templates to be added to this repo.
+        Delete an index
         """
-        return cast(dict[str, str], self.session.post("/index", json=pydantic_to_dict(index_def)).json())
+        if index_name:
+            index_id = self.get_index_manifest(index_name)["id"]
 
-    def update_index(self, index_def: Index) -> dict[str, str]:
-        """
-        Update (full object replacement) an index
-        """
-        return cast(dict[str, str], self.session.put(f"/index/{index_def.id}", json=pydantic_to_dict(index_def)).json())
+        return cast(HTTPResJson, self.session.delete(f"/index/{index_id}").json())
 
-    def patch_index(self, index_id: str, index_updates: Manifest) -> dict[str, str]:
+    @_name_or_id
+    def lock_index(self, index_name: str | None = None, index_id: str | None = None) -> HTTPResJson:
         """
-        Patch an index - index_updates is a partial index manifest
+        Lock an index.
+
+        Locks/unlocks must be done in isolated operations; ie you cannot lock a manifest while changing anything else,
+        similarly you cannot change anything else while unlocking a manifest.
+
+        Locking an index prevents all writes (PUT/PATCH/DELETE) on that index until it is explicitly unlocked.
         """
-        return cast(dict[str, str], self.session.patch(f"/index/{index_id}", json=index_updates).json())
+        model = self.get_index_model(index_name=index_name, index_id=index_id)
+
+        if (status := model.status).locked_after:
+            raise ValueError("Index is already locked")
+
+        status.locked_after = freezable_utcnow_iso()
+        payload = {"status": pydantic_to_dict(status)}
+        return cast(HTTPResJson, self.patch_index(index_id=model.id, updates=payload, auto_status=False))
 
-    def delete_index(self, index_id: str) -> dict[str, str]:
+    @_name_or_id
+    def unlock_index(self, index_name: str | None = None, index_id: str | None = None) -> HTTPResJson:
         """
-        Delete an index
+        Unlock an index; see "lock_index"
         """
-        return cast(dict[str, str], self.session.delete(f"/index/{index_id}").json())
+        model = self.get_index_model(index_name=index_name, index_id=index_id)
+
+        if (status := model.status).locked_after is None:
```

Comment:
Should this really be an exception? What's the harm in unlocking an index that is already unlocked? (just curious about the rationale. Don't need to change it)